pokedex/simulator/difftest/src/pokedex.rs

```diff
@@ -66,3 +61,60 @@ impl Display for InsnCommit {
         )
     }
 }
+
+impl crate::replay::IsCpuState for InsnCommit {
+    fn get_pc(&self) -> u32 {
+        self.pc as u32
+    }
+
+    fn has_reset(&self) -> Option<u32> {
+        self.find_reset_vector()
+    }
+
+    fn write_cpu_state(
+        &self,
+        state: &mut crate::replay::CpuState,
+    ) -> crate::replay::StateCheckType {
+        state.pc = self.pc as u32;
+
+        let mut check_ty = crate::replay::StateCheckType::default();
+
+        let mut has_frf_write = false;
+        self.states_changed.iter().for_each(|write| {
+            use crate::replay::CsrCheckType;
+            use ModelStateWrite::*;
+
+            match write {
+                Xrf { rd, value } => {
+                    if state.write_gpr((*rd) as usize, *value).is_some() {
+                        check_ty.gpr_rd = Some((*rd) as usize);
+                    }
+                }
+                Frf { rd, value } => {
+                    if state.write_fpr((*rd) as usize, *value).is_some() {
+                        check_ty.fpr_rd = Some((*rd) as usize);
+                    }
+                    has_frf_write = true;
+                    if matches!(check_ty.csr_mask, CsrCheckType::AllCsr) {
+                        check_ty.csr_mask = CsrCheckType::FpCsrOnly;
+                    }
+                }
+                Csr { idx, name, value } => {
+                    if state.write_csr(name, *idx, *value).is_some() {
+                        if has_frf_write {
+                            check_ty.csr_mask = crate::replay::CsrCheckType::FpCsrOnly;
+                        } else {
+                            check_ty.csr_mask = crate::replay::CsrCheckType::AllCsr;
+                        }
+                    }
+                }
```

Comment:
The logic to determine csr_mask looks too confusing. Consider adding a new `has_csr_write` local variable, then computing csr_mask from has_frf_write and has_csr_write at the very end?